core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -3189,11 +3192,54 @@ protected Scope visitSubqueryExpression(SubqueryExpression node, Optional<Scope>
         @Override
         protected Scope visitSetOperation(SetOperation node, Optional<Scope> scope)
         {
-            checkState(node.getRelations().size() == 2, "relations size must be 2");
+            List<Relation> relations = node.getRelations();
 
-            List<RelationType> childrenTypes = node.getRelations().stream()
-                    .map(relation -> process(relation, scope).getRelationType().withOnlyVisibleFields())
-                    .collect(toImmutableList());
+            checkState(relations.size() == 2, "relations size must be 2");
+
+            List<RelationType> childrenTypes = new ArrayList<>();
+            if (node.getCorresponding().isPresent()) {
+                Corresponding corresponding = node.getCorresponding().get();
+                if (!corresponding.getColumns().isEmpty()) {
+                    throw semanticException(NOT_SUPPORTED, node, "CORRESPONDING with columns is unsupported");
+                }
+
+                RelationType leftRelation = process(relations.getFirst(), scope).getRelationType().withOnlyVisibleFields();
+                RelationType rightRelation = process(relations.getLast(), scope).getRelationType().withOnlyVisibleFields();
+
+                Map<String, Integer> leftFieldsByName = buildNameToIndex(node, leftRelation);
+                Map<String, Integer> rightFieldsByName = buildNameToIndex(node, rightRelation);
+
+                List<String> correspondingColumns = leftFieldsByName.keySet().stream()
+                        .filter(rightFieldsByName::containsKey)
+                        .collect(toImmutableList());
+
+                if (correspondingColumns.isEmpty()) {
+                    throw semanticException(NOT_SUPPORTED, node, "No corresponding columns");
```

Comment:
Why NOT_SUPPORTED? This should be a proper error to indicate there are no corresponding columns, since it's an invalid query.

Reply:
Added a new `INVALID_CORRESPONDING_COLUMN` error code.

Reply:
I think you can use `MISMATCHED_COLUMN_ALIASES` instead of creating a new one.

Reply:
Thanks, changed to `MISMATCHED_COLUMN_ALIASES`.